### DISCUSSION TITLE
blacksheep/routing/

### DISCUSSION THREAD

**Discussion by giscus** [Neoteroi/BlackSheep-Docs#45] [Q&A]
# blacksheep/routing/

The official documentation website of BlackSheep, a fast ASGI web framework for Python asyncio.

https://www.neoteroi.dev/blacksheep/routing/

**Comment by codecakes** [comment]
is this possible or could it be possible?

```python3

app.controllers_router += [controllers_router1, controllers_router2,..]
```

**Reply by RobertoPrevato** [reply]
Apologies for the late reply. The exact code API of your example is not possible with the current code, but it would be sufficient to define a subclass of `RoutesRegistry` that yields routes defined in `[controllers_router1, controllers_router2,..]`, and set it as the application's `controllers_router` or directly to the instance of `Router`. 

It _should_ be sufficient to implemet the two methods below:

```python
class MyRoutesRegistry(RoutesRegistry):

    def __iter__(self): ...

    def reset(self): ...
```

Then if you really want to use the `+=` syntax of your example, you could implement the `__iadd__` protocol in your custom `RoutesRegistry`. Example (I didn't try):

```python
class MyRoutesRegistry(RoutesRegistry):

    def __iter__(self):
        # Your implementation here
        ...

    def reset(self):
        # Your implementation here
        ...

    def extend(self, route_registry):
        # Your implementation here
        ...

    def __iadd__(self, other):
        if isinstance(other, list):
            for router in other:
                if isinstance(router, RoutesRegistry):
                    self.extend(router)
                else:
                    raise TypeError("All items must be instances of RoutesRegistry")
        else:
            raise TypeError("Can only add a list of RoutesRegistry instances")
        return self
```

```python
controllers_router1 = MyRoutesRegistry()
controllers_router2 = MyRoutesRegistry()
controllers_router3 = MyRoutesRegistry()

app.controllers_router += [controllers_router1, controllers_router2, controllers_router3]
```
